dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -639,57 +611,106 @@ void DMFileReader::readFromDiskOrSharingCache(
+        getColumnFromCache(
+            data_sharing_col_data_cache,
+            cd,
+            start_pack_id,
+            pack_count,
+            read_rows,
+            mutable_col,
+            [&](const ColumnDefine & cd, size_t start_pack_id, size_t, size_t read_rows) {
+                // If there are concurrent read requests, this data is likely to be shared.
+                // So the allocation and deallocation of this data may not be in the same MemoryTracker.
+                // This can lead to inaccurate memory statistics of MemoryTracker.
+                // To solve this problem, we use a independent global memory tracker to trace the shared column data in the data_sharing_col_data_cache.
+                MemoryTrackerSetter mem_tracker_guard(true, nullptr);
+                return readFromDisk(cd, start_pack_id, read_rows);
+            });
+        column = std::move(mutable_col);
+    }
+    else
+    {
+        column = readFromDisk(cd, start_pack_id, read_rows);
     }
 
-    if (enable_sharing_column && col_data_cache != nullptr)
+    // Set the column to DMFileReaderPool to share the column data.
+    if (enable_sharing_column && data_sharing_col_data_cache != nullptr)
     {
-        DMFileReaderPool::instance().set(*this, column_define.id, start_pack_id, pack_count, column);
+        DMFileReaderPool::instance().set(*this, cd.id, start_pack_id, pack_count, column);
+        // Delete column from local cache since it is not used anymore.
+        data_sharing_col_data_cache->delColumn(cd.id, next_pack_id);
     }
+    return column;
 }
 
-void DMFileReader::addCachedPacks(ColId col_id, size_t start_pack_id, size_t pack_count, ColumnPtr & col) const
+void DMFileReader::addColumnToCache(
+    ColumnCachePtr data_cache,
+    ColId col_id,
+    size_t start_pack_id,
+    size_t pack_count,
+    ColumnPtr & col)
 {
-    if (col_data_cache == nullptr)
-    {
+    if (data_cache == nullptr)
         return;
-    }
-    if (next_pack_id >= start_pack_id + pack_count)
-    {
-        col_data_cache->addStale();
-    }
-    else
+
+    const auto & pack_stats = dmfile->getPackStats();
+    size_t rows_offset = 0;
+    for (size_t cursor = start_pack_id; cursor < start_pack_id + pack_count; ++cursor)
     {
-        col_data_cache->add(col_id, start_pack_id, pack_count, col);
+        data_cache->tryPutColumn(cursor, col_id, col, rows_offset, pack_stats[cursor].rows);
+        rows_offset += pack_stats[cursor].rows;
     }
 }
 
-bool DMFileReader::getCachedPacks(
-    ColId col_id,
+void DMFileReader::getColumnFromCache(
+    ColumnCachePtr data_cache,
+    const ColumnDefine & cd,
     size_t start_pack_id,
     size_t pack_count,
     size_t read_rows,
-    ColumnPtr & col) const
+    MutableColumnPtr & result_col,
+    // column_define, column_id, start_pack_id, pack_count, read_rows
+    std::function<ColumnPtr(const ColumnDefine &, size_t, size_t, size_t)> on_cache_miss)
```

Comment:
Suggest replacing the param `MutableColumnPtr & result_col` by `const DataTypePtr & type_on_disk` and the return type to `ColumnPtr`. So that we can:
* `auto mutable_col = type_on_disk->createColumn(read_rows);` instead of creating an empty col then reallocate the size of the column
* Return the column to simplify the caller code

Reply:
done